### PR TITLE
WT-8534 Allow retrieving the checkpoint snapshot for backup restore recovery

### PR DESCRIPTION
**Pull request by kommiharibabu**
WiredTiger backup doesn't contains the WiredTiger.turtle file to
identify the source WiredTiger version of the backup. The saved
checkpoint is retrieved as part of the backup restore recovery
to fix any inconsistencies that exist.

### CODE REVIEW

File: src/txn/txn_recover.c [hunk 1]
```diff
@@ -484,9 +484,10 @@ __recovery_set_checkpoint_snapshot(WT_SESSION_IMPL *session)
 
     /*
      * WiredTiger versions 10.0.1 onward have a valid checkpoint snapshot on-disk. Ignore reading
-     * the on-disk checkpoint snapshot from older versions.
+     * the on-disk checkpoint snapshot from older versions. WiredTiger backup doesn't contain the
+     * WiredTiger.turtle file. Allow retrieving the saved checkpoint snapshot if it exists.
      */
-    if (conn->recovery_major < 10 ||
+    if ((conn->recovery_major != 0 && conn->recovery_major < 10) ||
```
Comment (agorrod) [inline src/txn/txn_recover.c:490]: I believe this change is right, but it makes me uncomfortable because it relies on the explicit 0 initialization of the WT_CONNECTION structure, and that not being overwritten in an unrelated `conn_version_verify` function.

Could you also update `conn_version_verify` to deal explicitly with the case of having no turtle file. That might just be a comment update, but it could also be an explicit setting of the versions to no-op values.

Comment (kommiharibabu) [reply src/txn/txn_recover.c:495]: Initialised the variables in the `conn_version_verify` function to avoid the dependency of connection structure being initialised to 0.

File: src/conn/conn_api.c [hunk 2]
```diff
@@ -2505,6 +2505,14 @@ __conn_version_verify(WT_SESSION_IMPL *session)
 
     conn = S2C(session);
 
+    /*
+     * Initialize the recovery version variables. Later these will be updated with proper WiredTiger
+     * version if the database is being opened for recovery.
```
Comment (agorrod) [inline src/conn/conn_api.c:2510]: I think it's worth updating this comment to be more explicit about what "for recovery" means. Maybe even referencing backup

Comment (kommiharibabu) [reply src/conn/conn_api.c]: Thanks. I try to expand the comment a bit with details of recovery. Please check and let me know your comments?

File: src/conn/conn_api.c [hunk 3]
```diff
@@ -2505,6 +2505,16 @@ __conn_version_verify(WT_SESSION_IMPL *session)
...
 
+    /*
+     * Initialize the recovery version variables. Later these will be updated with proper WiredTiger
+     * version if the database is being opened for recovery from the WiredTiger.turtle file.
+     * WiredTiger recovery occurs not only when the database is restarted after a clean or unclean
+     * shutdown, but also when the backup is restored.
+     */
+    conn->recovery_major = 0;
+    conn->recovery_minor = 0;
+    conn->recovery_patch = 0;
```
Comment (sueloverso) [inline src/conn/conn_api.c:2516]: This change is unnecessary. The connection structure is guaranteed to be zeroed on allocation. There is a lot of code that depends on this. Having this here is confusing because it causes me to wonder what they were set to that this is needed/here. I suggest removing this.

File: src/txn/txn_recover.c (hunk 1 above)
Comment (sueloverso) [reply src/txn/txn_recover.c:495]: Heh - and I just left a big comment about the addition of that (it shows earlier in the diff so I saw that before opening this resolved comment). @agorrod I find initializing a few fields in the conn structure to 0 more confusing when it is only called early in wiredtiger_open. That makes me have to dig around to wonder why it wouldn't be 0 from the calloc. There's a lot of code that depends on the connection being zeroed on allocation. We don't initialize every other field explicitly in random places.

File: test/suite/test_checkpoint_snapshot02.py [hunk 4]
```diff
@@ -42,17 +42,24 @@ class test_checkpoint_snapshot02(wttest.WiredTigerTestCase):
...
+    restart_values = [
+        ("crash_restart", dict(restart=True)),
+        ("backup", dict(restart=False)),
+    ]
+
+    scenarios = make_scenarios(format_values, restart_values)
 
     def conn_config(self):
-        config = 'cache_size=10MB,statistics=(all),statistics_log=(json,on_close,wait=1),log=(enabled=true),timing_stress_for_test=[checkpoint_slow]'
+        config = 'cache_size=10MB,statistics=(all),statistics_log=(json,on_close,wait=1),log=(archive=false,enabled=true),timing_stress_for_test=[checkpoint_slow]'
```
Comment (sueloverso) [inline test/suite/test_checkpoint_snapshot02.py:62]: Why did you add this to turn off archiving? That is not needed for backup.

**Review by sueloverso — APPROVED** [review]
LGTM with a few minor comments.

File: src/conn/conn_api.c (hunk 3 above)
Comment (agorrod) [reply src/conn/conn_api.c:2516]: I aksed for something like this. I had to trace through the code to get certainty that it'll work. I'm OK to have a different solution, but the implicit relationship where not retrieving a version when the turtle file doesn't exist implies a 0 release, and we rely on that for future decision making feels fragile.

**Review by agorrod — APPROVED** [review]
lgtm

File: src/conn/conn_api.c (hunk 3 above)
Comment (sueloverso) [reply src/conn/conn_api.c:2510]: It is only looked at in recovery code so we know it is during wiredtiger_open where we've allocated a connection. Maybe a comment in the recovery code saying "this was set up when we read the turtle file in <function>." and then in that function add a comment that it is set here or will be zero if there is no turtle file from a backup (or whatever the details of when we do/do not are).

**Review by sulabhM — APPROVED** [review]

File: test/suite/test_checkpoint_snapshot02.py [hunk 5]
```diff
@@ -66,6 +73,22 @@ def moresetup(self):
...
+        self.pr('Full backup from '+ fromdir + ' to ' + todir + ': ')
+        os.mkdir(todir)
+        while True:
+            ret = cursor.next()
+            if ret != 0:
+                break
+            bkup_file = cursor.get_key()
+            sz = os.path.getsize(fromdir + "/" + bkup_file)
+            self.pr('Copy from: ' + bkup_file + ' (' + str(sz) + ') to ' + todir)
+            shutil.copy(fromdir + "/" + bkup_file, todir)
```
Comment (AlexeyAnisimovAU) [inline test/suite/test_checkpoint_snapshot02.py:88]: It's better to use `os.path.join()` instead of concatenating the file path manually.
Also, we do it twice on lines 86 and 88. It's would be better if we avoid duplication.

File: test/suite/test_checkpoint_snapshot02.py [hunk 6]
```diff
@@ -264,7 +302,13 @@ def test_checkpoint_snapshot_with_txnid_and_timestamp(self):
         self.assertGreater(inconsistent_ckpt, 0)
         self.assertGreaterEqual(keys_removed, 0)
 
-        simulate_crash_restart(self, "RESTART", "RESTART2")
+        if self.restart == True:
+            simulate_crash_restart(self, "RESTART", "RESTART2")
+        else:
+            #Take a backup and restore it.
+            self.take_full_backup(self.backup_dir, self.backup_dir2)
+            self.reopen_conn(self.backup_dir2)
```
Comment (AlexeyAnisimovAU) [inline test/suite/test_checkpoint_snapshot02.py:310]: Can we extract the if/else block above into a separate method to avoid duplication?

File: src/txn/txn_recover.c [hunk 7]
```diff
@@ -482,11 +482,17 @@ __recovery_set_checkpoint_snapshot(WT_SESSION_IMPL *session)
...
+    conn->recovery_ckpt_snap_min = WT_TXN_NONE;
+    conn->recovery_ckpt_snap_max = WT_TXN_NONE;
+    conn->recovery_ckpt_snapshot_count = 0;
+
     /*
      * WiredTiger versions 10.0.1 onward have a valid checkpoint snapshot on-disk. Ignore reading
-     * the on-disk checkpoint snapshot from older versions.
+     * the on-disk checkpoint snapshot from older versions. WiredTiger backup doesn't contain the
+     * WiredTiger.turtle file. Allow retrieving the saved checkpoint snapshot if it exists.
      */
```
Comment (AlexeyAnisimovAU) [inline src/txn/txn_recover.c:494]: Do you think it would be safer if we add `conn->recovery_major == 0` to this check as well?
https://github.com/wiredtiger/wiredtiger/blob/67e77c4fe970124e2c343c68596e9d39db265fae/src/txn/txn_recover.c#L990 
Something like this maybe?
`if (conn->unclean_shutdown && (conn->recovery_major == 0 || conn->recovery_major == 10) && conn->recovery_minor == 0 && conn->recovery_patch == 0)`

File: src/conn/conn_api.c (hunk 3 above)
Comment (AlexeyAnisimovAU) [reply src/conn/conn_api.c:2510]: As an alternative, we can initialize those variables in all the places where we don't call `__wt_turtle_validate_version`.

**Review by AlexeyAnisimovAU — APPROVED** [review]
LGTM, but I have a couple of notes

File: src/txn/txn_recover.c (hunk 7 above)
Comment (agorrod) [reply src/txn/txn_recover.c:499]: I don't think that's necessary. Other WiredTiger major versions never generated a checkpoint snapshot. I've pushed a comment that I think covers this case.

File: src/conn/conn_api.c (hunk 3 above)
Comment (kommiharibabu) [reply src/conn/conn_api.c:2510]: `__conn_version_verify` is the first function that checks for the turtle file, I feel initializing it here may be enough.

File: test/suite/test_checkpoint_snapshot02.py (hunk 4 above)
Comment (kommiharibabu) [reply test/suite/test_checkpoint_snapshot02.py]: Fixed. Thanks.

File: test/suite/test_checkpoint_snapshot02.py (hunk 5 above)
Comment (kommiharibabu) [reply test/suite/test_checkpoint_snapshot02.py]: Fixed. Thanks.

File: test/suite/test_checkpoint_snapshot02.py (hunk 6 above)
Comment (kommiharibabu) [reply test/suite/test_checkpoint_snapshot02.py]: Added a new function to remove duplicate code.
